### DISCUSSION TITLE
How to use extensions such as FairyGUI and DragonBones with Axmol?

### DISCUSSION THREAD

**Discussion by awetos** [axmolengine/axmol#1348] [Q&A]
Hi, 
I think I fundamentally do not understand linking a library/using extensions and using them in C++. I have been using Unity and C#, and I have never used Cocos2d-x. But I wanted to learn it because it is open source and offers many of the same functionalities I want to use in my game. 

### I created a file in FairyGUI with cocos2d-x option. 
![image](https://github.com/axmolengine/axmol/assets/78708899/e181b1ac-2c05-4c91-890e-ec66e6f291c6)

### I put it in this folder (Is this the wrong folder?)
![image](https://github.com/axmolengine/axmol/assets/78708899/ca278390-b8ff-4f26-9f13-03050e3d33c7)

### I see in the official FairyGUI demo they add this include
![image](https://github.com/axmolengine/axmol/assets/78708899/52e9816f-9c8f-4671-90e7-b7b664149350)

### So I add that to my HelloCpp
![image](https://github.com/axmolengine/axmol/assets/78708899/64fd61e1-6138-4ae3-9cc6-97ab51a569d8)

### Then I see they create an object with GROOT; but nothing happens when I type FairyGUI, GROOT, or UIPackage.
![image](https://github.com/axmolengine/axmol/assets/78708899/8127b57a-142e-4de5-a6d4-a090dc63ff96)

### I see that there are already all kinds of extensions in the library
![image](https://github.com/axmolengine/axmol/assets/78708899/f0511276-caef-4790-a69b-1ddc0e57abc1)

### I am also not sure if I am opening the right project file. 
![image](https://github.com/axmolengine/axmol/assets/78708899/0db9e44c-d619-422c-9645-016d92d4cd67)
### Is it this one?

**Sorry for the very basic question**. I hope you can answer it for anyone else who is new and wanting to use Axmol. 
If there are other resources for learning to use Axmol, C++, Cocos2d-x, please let me know. I have watched some videos on YouTube for Cocos2d-x. I know how to do some programming but I don't know how to setup an engine like this.
Thank you.

**Comment by rh101** [comment]
If you're not already familiar with programming in C++, then that is where you need to start, and diving into a C++ game engine without that fundamental knowledge is most likely going to waste your time and cause you a lot of frustration.  If you do need to get into C++, then there are millions of resources available to you, but this is not the right place to be asking those types of questions.

If you are familiar with C++, and want help with this game engine, then sure, ask away.  In your particular case, the actual file you need to open in Visual Studio is the solution file, with extension "sln", being `HelloCpp.sln`.  

If you follow the [documentation](https://github.com/axmolengine/axmol/blob/dev/CMakeOptions.md), you would see which settings are required to enable or disable certain "extensions", which are just libraries that link with the engine.  If you don't need the functionality in those libraries, then disabling them will reduce the size of the final executable and also reduce the time it takes to compile your application.

For example, if you want to use DragonBones, then the setting is `AX_ENABLE_EXT_DRAGONBONES`.  If you search for that in the code (specifically the cmake files `CMakeLists.txt` and `*.cmake`), you would find this:

`option(AX_ENABLE_EXT_DRAGONBONES "Build extension DragonBones" ON)`

It's enabled by default.

You can control if it is enabled or disabled in your own project `CMakeLists.txt`, for example, this is how you would disable it (with `OFF`):

`set(AX_ENABLE_EXT_DRAGONBONES OFF CACHE BOOL "Build extension DragonBones" FORCE)`

That needs to go above the line that has `include(AXBuildSet)` in your `CMakeLists.txt`. For example:

```
...
cmake_minimum_required(VERSION 3.10)

set(APP_NAME HelloCpp)

project(${APP_NAME})

# Enable/Disable extensions here
set(AX_ENABLE_EXT_DRAGONBONES OFF CACHE BOOL "Build extension DragonBones" FORCE)

if(NOT DEFINED BUILD_ENGINE_DONE) # to test HelloCpp into root project
    if(XCODE)
        set(CMAKE_XCODE_GENERATE_TOP_LEVEL_PROJECT_ONLY TRUE)
    endif()
...
```

In order to use these libraries, it's best to refer to the cpp_tests project.  Compile it, run it, see how things work, and then read the code to figure out how things are done. It's all in there.

**Comment by rh101** [comment]
A few reasonable Cocos2d-x tutorials that would also apply to Axmol:

[here](https://www.kodeco.com/1848-cocos2d-x-tutorial-for-beginners)

and

[here](http://wizardfu.com/book/cocos2d-x/) - this also has a section on `C++`, with more links that may help you.

Just disregard any sections on installing and setting up, since they may not apply to Axmol.  For that info refer to [this](https://github.com/axmolengine/axmol/blob/dev/docs/DevSetup.md).

Any guide to cocos2d-x will mostly apply to Axmol too. For any differences, refer to the `cpp_tests` project in Axmol source tree.

Don't forget to check [this post](https://github.com/axmolengine/axmol/discussions/694) for links that would also help.

**Comment by awetos** [comment]
Thank you for the resources
I will study c++, compiling, and using libraries and read the engine's documentation to make sure my questions are about the engine next time

**Reply by DelinWorks** [reply]
If you get stuck at anything any time soon, you can give axmol's discord server a visit! it will be more straightforward there. https://discord.gg/PTJUwmuW

**Reply by danialias** [reply]
Also, you don't need to use FairyGUI or any extra extension to start working with Axmol. First you should try the basics: draw a sprite, move it, add a background, and some sounds. Then you can add another scene, and keep improving your app from there.
